sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -452,6 +452,17 @@ class Analyzer(
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
+        // A partitioned relation's schema can be different from the input logicalPlan, since
+        // partition columns are all moved after data columns. We Project to adjust the ordering.
+        val input = if (parts.nonEmpty) {
+          val (inputPartCols, inputDataCols) = child.output.partition { attr =>
+            parts.contains(attr.name)
+          }
+          Project(inputDataCols ++ inputPartCols, child)
+        } else {
+          child
+        }
```

Comment:
Looks like Hive uses ordering not name to take dynamic partition columns. I am not sure if we want to completely follow this Hive behavior. DataFrameWriter's insertInto doesn't follow this. Besides, the rule in Analyzer is not completely follow this too.

@liancheng @rxin @cloud-fan What do you think? Do you think we should change current behavior to follow Hive?